File: sistemas/java/DecomposicaoLU.java
package exatos;

import java.math.BigDecimal;

public class DecomposicaoLU {

	public static void lu(double a[][]) {

		double lower[][] = new double[a[0].length][a.length];
		double upper[][] = new double[a[0].length][a.length];

		System.out.println("Matriz A");
		for (int i = 0; i < a[0].length; i++) {
			for (int j = 0; j < a.length; j++) {
				System.out.print(a[i][j] + " ");
			}
			System.out.println();
		}

		int n = a[0].length;

		System.out.println("N = " + n);

		for (int j = 0; j < n; j++) {
			lower[j][j] = 1;
			for (int i = 0; i <= j; i++) {
				double sum = BigDecimal.ZERO.doubleValue();
				for (int k = 0; k < i; k++) {
					sum += upper[k][j] * lower[i][k];
				}
				upper[i][j] = a[i][j] - sum;
			}

			for (int i = j; i < n; i++) {
				double sum = BigDecimal.ZERO.doubleValue();
				for (int k = 0; k < i; k++) {
					sum += upper[k][j] * lower[i][k];
				}
				lower[i][j] = (a[i][j] - sum) / upper[j][j];
			}
		}

		System.out.println();
		System.out.println("Matriz L");
		for (int i = 0; i < lower[0].length; i++) {
			for (int j = 0; j < lower.length; j++) {
				System.out.print(lower[i][j] + " ");
			}
			System.out.println();
		}

		System.out.println();
		System.out.println("Matriz U");
		for (int i = 0; i < upper[0].length; i++) {
			for (int j = 0; j < upper.length; j++) {
				System.out.print(upper[i][j] + " ");
			}
			System.out.println();
		}
	}

	public static void main(String[] args) {
		double a[][] = new double[][] { { 2, -1, -2 }, { -4, 6, 3 }, { -4, -2, 8 } };
		lu(a);
	}

}

File: outros/aps/decomposicao_lu.cpp
#include "lu.h"

void print_matrix(const vvd &matriz, const string &title) {
    cout << title << '\n';
    int i, j;
    loop(i, 0, matriz[0].size()) {
        loop(j, 0, matriz[i].size()) {
            cout << matriz[i][j] << '\t';
        }
        cout << endl;
    }
}

double sum(const vvd &a, const vvd &b, int j, int i) {
    double sum = 0.0;
    int k;
    loop(k, 0, i) {
        sum += a[k][j] * b[i][k];
    }
    return sum;
}

void lu(const vvd &matriz) {
    int n = matriz[0].size();

    vvd lower = vvd(n, vector<double>(n));
    vvd upper = vvd(n, vector<double>(n));

    print_matrix(matriz, "Matriz A");

    loop(j, 0, n) {
        lower[j][j] = 1;
        loop_eq(i, 0, j) {
            upper[i][j] = matriz[i][j] - sum(upper, lower, j, i);
        }

        loop(i, j, n) {
            lower[i][j] = (matriz[i][j] - sum(upper, lower, j, i)) / upper[j][j];
        }
    }
    
    cout << '\n';
    print_matrix(lower, "Matriz L");

    cout << '\n';
    print_matrix(upper, "Matriz U");
}